File: backend-go/internal/httpapi/sharing_handlers.go
package httpapi

import (
	"net/http"

	"github.com/danielraab/na-notes-app/backend-go/internal/notes"
)

func (d *Deps) handleListShares(w http.ResponseWriter, r *http.Request, userID string) {
	noteID := r.PathValue("noteId")
	shares, public, err := d.Notes.ListShares(r.Context(), noteID, userID)
	if err != nil {
		respondDomainError(w, err)
		return
	}

	userShares := make([]userShareDTO, 0, len(shares))
	for _, s := range shares {
		userShares = append(userShares, toUserShareDTO(s))
	}

	var publicDTO *publicShareDTO
	if public != nil {
		dto := toPublicShareDTO(*public, d.Config.FrontendURL+"/shared/"+public.Token)
		publicDTO = &dto
	}

	respondJSON(w, http.StatusOK, struct {
		UserShares  []userShareDTO  `json:"userShares"`
		PublicShare *publicShareDTO `json:"publicShare"`
	}{UserShares: userShares, PublicShare: publicDTO})
}

func (d *Deps) handleCreateShare(w http.ResponseWriter, r *http.Request, userID string) {
	noteID := r.PathValue("noteId")

	var in struct {
		UserID     string `json:"userId"`
		Permission string `json:"permission"`
	}
	if err := decodeJSON(r, &in); err != nil {
		respondError(w, http.StatusBadRequest, "VALIDATION_ERROR", "invalid request body")
		return
	}
	if in.Permission != string(notes.PermissionRead) && in.Permission != string(notes.PermissionEdit) {
		respondError(w, http.StatusBadRequest, "VALIDATION_ERROR", "permission must be 'read' or 'edit'")
		return
	}

	share, err := d.Notes.ShareWithUser(r.Context(), noteID, userID, in.UserID, notes.Permission(in.Permission))
	if err != nil {
		respondDomainError(w, err)
		return
	}
	respondJSON(w, http.StatusCreated, toUserShareDTO(share))
}

func (d *Deps) handleDeleteShare(w http.ResponseWriter, r *http.Request, userID string) {
	noteID := r.PathValue("noteId")
	targetUserID := r.PathValue("userId")
	if err := d.Notes.RevokeShare(r.Context(), noteID, userID, targetUserID); err != nil {
		respondDomainError(w, err)
		return
	}
	respondNoContent(w)
}

func (d *Deps) handleCreatePublicShare(w http.ResponseWriter, r *http.Request, userID string) {
	noteID := r.PathValue("noteId")
	ps, url, err := d.Notes.CreatePublicShare(r.Context(), noteID, userID)
	if err != nil {
		respondDomainError(w, err)
		return
	}
	respondJSON(w, http.StatusCreated, toPublicShareDTO(ps, url))
}

func (d *Deps) handleDeletePublicShare(w http.ResponseWriter, r *http.Request, userID string) {
	noteID := r.PathValue("noteId")
	if err := d.Notes.RevokePublicShare(r.Context(), noteID, userID); err != nil {
		respondDomainError(w, err)
		return
	}
	respondNoContent(w)
}

File: backend-rust/src/httpapi/sharing_handlers.rs
use std::sync::Arc;

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;

use crate::notes::Permission;

use super::dto::{to_public_share_dto, to_user_share_dto, CreateShareDto, NoteSharesDto};
use super::middleware::RequireAuth;
use super::respond::ApiError;
use super::server::Deps;

pub async fn handle_list_shares(
    State(deps): State<Arc<Deps>>,
    RequireAuth(user_id): RequireAuth,
    Path(note_id): Path<String>,
) -> Result<Response, ApiError> {
    let (shares, public) = deps.notes.list_shares(&note_id, &user_id).await?;
    let user_shares = shares.into_iter().map(to_user_share_dto).collect();
    let public_share = public.map(|ps| {
        let url = format!("{}/shared/{}", deps.config.frontend_url, ps.token);
        to_public_share_dto(ps, url)
    });
    Ok(Json(NoteSharesDto {
        user_shares,
        public_share,
    })
    .into_response())
}

pub async fn handle_create_share(
    State(deps): State<Arc<Deps>>,
    RequireAuth(user_id): RequireAuth,
    Path(note_id): Path<String>,
    body: Result<Json<CreateShareDto>, axum::extract::rejection::JsonRejection>,
) -> Result<Response, ApiError> {
    let Json(input) = body.map_err(|_| {
        ApiError::new(
            StatusCode::BAD_REQUEST,
            "VALIDATION_ERROR",
            "invalid request body",
        )
    })?;

    let permission = match input.permission.as_str() {
        "read" => Permission::Read,
        "edit" => Permission::Edit,
        _ => {
            return Err(ApiError::new(
                StatusCode::BAD_REQUEST,
                "VALIDATION_ERROR",
                "permission must be 'read' or 'edit'",
            ))
        }
    };

    let share = deps
        .notes
        .share_with_user(&note_id, &user_id, &input.user_id, permission)
        .await?;
    Ok((StatusCode::CREATED, Json(to_user_share_dto(share))).into_response())
}

pub async fn handle_delete_share(
    State(deps): State<Arc<Deps>>,
    RequireAuth(user_id): RequireAuth,
    Path((note_id, target_user_id)): Path<(String, String)>,
) -> Result<StatusCode, ApiError> {
    deps.notes
        .revoke_share(&note_id, &user_id, &target_user_id)
        .await?;
    Ok(StatusCode::NO_CONTENT)
}

pub async fn handle_create_public_share(
    State(deps): State<Arc<Deps>>,
    RequireAuth(user_id): RequireAuth,
    Path(note_id): Path<String>,
) -> Result<Response, ApiError> {
    let (ps, url) = deps.notes.create_public_share(&note_id, &user_id).await?;
    Ok((StatusCode::CREATED, Json(to_public_share_dto(ps, url))).into_response())
}

pub async fn handle_delete_public_share(
    State(deps): State<Arc<Deps>>,
    RequireAuth(user_id): RequireAuth,
    Path(note_id): Path<String>,
) -> Result<StatusCode, ApiError> {
    deps.notes.revoke_public_share(&note_id, &user_id).await?;
    Ok(StatusCode::NO_CONTENT)
}
